src/main/java/org/motechproject/prevac/service/impl/PrimeVaccinationScheduleServiceImpl.java

```diff
@@ -55,47 +55,78 @@
 
     @Override
     public PrimeVaccinationScheduleDto createOrUpdateWithDto(PrimeVaccinationScheduleDto dto, Boolean ignoreLimitation) {
-        Visit primeDetails = visitBookingDetailsDataService.findById(dto.getVisitBookingDetailsId());
-        Visit screeningDetails = getScreeningDetails(primeDetails);
+        validateDate(dto);
 
-        if (primeDetails == null || screeningDetails == null) {
-            throw new IllegalArgumentException("Cannot save, because details for Visit not found");
+        Visit primeVisit = visitBookingDetailsDataService.findById(dto.getVisitBookingDetailsId());
+        // We have an update
+        if (primeVisit != null) {
+            Visit screeningVisit = getScreeningDetails(primeVisit);
+            if (screeningVisit == null) {
+                throw new IllegalArgumentException("Cannot save, because details for Visit not found");
+            }
+            return new PrimeVaccinationScheduleDto(updateVisitWithDto(primeVisit, screeningVisit, dto));
         }
 
-        Clinic clinic = primeDetails.getClinic();
+        Subject subject = subjectDataService.findBySubjectId(dto.getParticipantId());
 
-        validateDate(dto);
+        List<Visit> visits = subject.getVisits();
```

Comment:
Where this list is used?

Reply:
It isn't used or returned anywhere but for some reason without this list screening isn't created correctly.

Reply:
What's the reason?

Reply:
I am assuming this is needed due to datanucleus caching magic. @pmuchowski feel free to chime in once you're ready